src/platform/i386/ioapic.c

```diff
@@ -0,0 +1,205 @@
+ioapic_findbyid(int id)
+{
+	unsigned int i = 0;
+
+	for (; i < ioapic_count; i++) {
+		if (id == (int)(ioapicinfo[i].ioapicid)) return &ioapicinfo[i];
+	}
+
+	return NULL;
+}
+
+void
+ioapic_int_override(struct intsrcovrride_cntl *iso)
+{
+	assert(iso->header.len == sizeof(struct intsrcovrride_cntl));
+
+	if (iso->source != iso->glb_int_num_off) {
+		struct ioapic_info *ioap = ioapic_findbygsi(iso->glb_int_num_off);
+
+		assert(ioap);
+		printk("\tInterrupt Source Override for [%u] => %u with IOAPIC %d\n", iso->source, iso->glb_int_num_off, ioap->ioapicid);
+		ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(iso->glb_int_num_off), iso->source + HW_IRQ_START);
+	}
+}
+
+void
+ioapic_int_mask(int irqnum)
+{
+	struct ioapic_info *ioap = ioapic_findbygsi(irqnum);
+
+	assert(ioap);
+	ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum), IOAPIC_INT_MASKED | irqnum);
+}
+
+void
+ioapic_int_unmask(int irqnum)
+{
+	struct ioapic_info *ioap = ioapic_findbygsi(irqnum);
+	u32_t val = 0;
+
+	assert(ioap);
+	val = ioapic_reg_read(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum));
+	ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum), val & ~IOAPIC_INT_MASKED);
+}
+
+void
+ioapic_int_enable(int irqnum, int cpunum, int addflag)
+{
+	struct ioapic_info *ioap = ioapic_findbygsi(irqnum);
+
+	assert(ioap);
+	if (addflag) {
+		/* TODO: logical destination = 1 and add core no or lapic number? */
+		assert(0);
+	} else {
+		ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum), irqnum + HW_IRQ_START);
+		ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum)+1, cpunum<<24);
+	}
+}
+
+void
+ioapic_int_disable(int irqnum)
+{
+	struct ioapic_info *ioap = ioapic_findbygsi(irqnum);
+
+	assert(ioap);
+	ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum), IOAPIC_INT_MASKED | irqnum);
+	ioapic_reg_write(ioap, IOAPIC_IOREDTBL_OFFSET(irqnum)+1, 0);
+}
+
+void
+ioapic_iter(struct ioapic_cntl *io)
+{
+	u32_t ver;
+	int ioent, j;
+	static int more = 0;
+
+	assert(io);
+
+	if (ioapic_count == IOAPIC_MAX) {
+		more ++;
+		printk("\t%d more than %d IOAPICs present..\n", more, IOAPIC_MAX);
+
+		return;
+	}
+	
+	ioapicinfo[ioapic_count].io_vaddr = (volatile void *)(io->ioapic_phys_addr);	
+	ioapicinfo[ioapic_count].ioapicid = io->ioapic_id;
+	ioapic_set_page(&(ioapicinfo[ioapic_count]), vm_set_supage((u32_t)(ioapicinfo[ioapic_count].io_vaddr)));
+
+	ver   = ioapic_reg_read(&ioapicinfo[ioapic_count], IOAPIC_IOAPICVER);
+	ioent = ((ver >> 16) & 0xFF) + 1;
+	printk("\tIOAPIC %d (counter:%d): Number of entries = %d\n", io->ioapic_id, ioapic_count, ioent);
+
+	ioapicinfo[ioapic_count].nentries    = ioent;
+	ioapicinfo[ioapic_count].glbint_base = io->glb_int_num_off;
+	ioapic_count ++;
+
+	for (j = 0; j < ioent; j++) ioapic_int_enable(io->glb_int_num_off + j, 0, 0); /* TODO: assign to different cores */
+}
+
+void
+ioapic_init(void)
+{
+	assert(ioapic_count);
+	pic_disable();
+
+	printk("Setting up IOAPIC (disabling PIC)\n");
+
+	/*
+	 * PCI Interrupts may need some attention here.
```

Comment:
If you have any documentation that shows why PCI might be complicated, you might want to throw a pointer to it in here.

Reply:
I'll look into that again. 
But I ported IOAPIC to RK yesterday and ran it on Qemu, it just worked fine. There is some debugging to do with HW however, I'm not able to get `ping` or `udpserver` to work with HW.

Reply:
The problem I saw in HW could be related to ACKs and not PCI. Will debug that for more info.
At the very least, I can see that the RK detects the NIC and assigns static IP as normal both on HW/Qemu.